### ISSUE TITLE
TPM error during boot

### ISSUE THREAD

**Issue by redragonx** [fox-it/linux-luks-tpm-boot#7]
So I followed the instructions and rebooted. However, I get this error below

`Tspi_Context_Connect failed: 0x00003011 - layer=tsp, code=0011`

Ubuntu 16.04.2
tpm ver 1.2

Yes, I can read tpm nvram during normal operations (after failing back to my passphrase. But unlocking fails.

Any ideas?

**Comment by morbitzer** [comment]
It seems that the tcsd daemon is not running in your initrd. If you append `break=init` to the kernel boot parameter in GRUB, the system will stop in the initrd. 
Could you then try to manually execute `getsecret.sh`? If you get the connection error, see if the tcsd daemon is running with `ps aux | grep tcsd` and try to start it by simply running `tcsd`. Then try to run `getsecret.sh` once more.

**Comment by redragonx** [comment]
Thanks for the info, I'll try it next week. 👍 

Do u know why the daemon is not starting at boot?

**Comment by morbitzer** [comment]
There could be multiple problems here: 

- **The daemon is not started in the initrd.** In this case, I would hope that you get an error message when trying to start it manually, which should help us to pin down the problem 
- **The daemon is started, but at the time getsecret.sh is invoked, stopped again.** For some reason, the tcsd daemon in the initrd stops after a single `tpm_*` command is issued - afterwards, it has to be restarted again. This only happens in the initrd, and I was not yet able to figure out why. (any tips appreciated!) 
So there might also be something going wrong in `getsecret.sh`, which issues two tpm-commands (the second one to stop the secret from being read again). Therefore it could also help to comment the two last lines of `getsecret.sh`. This way, the reading the secret after boot won't be blocked, but the normal bootup should work. 
- **The tcsd daemon is running normally in the initrd, and there is a different problem.** In that case, we'll have to think about other possibilities that could cause your issues.

**Comment by JorritVrieze** [comment]
I have the same issue as redragonx.  Ubuntu 16.04.2 and TPM 1.2 (also in 14.04.5)
In normal operations everything works fine but when booting not.
After rebooting i get the cryptsetup message 'bad password or options'.
In busybox v1.22.1 i did the ps aux | grep tcsd and it is running.
Also in busybox i tried tpm_nvinfo -l debug
Output:
Tspi_Context_Create_Success
Tspi_Context_Connect failed: 0x00003011 - layer=tsp, code=0011 (17), communication failure
Tspi_Context_FreeMemory success
Tspi_Context_Close success

Also tried to start the getsecret.sh manual but got the same error.

From the tpm-script i also get to message: chown: not found.

Hope you can help.

**Comment by morbitzer** [comment]
Did you make sure the `tcsd` daemon is running before executing the `tpm_nvinfo` command? (in busybox, after executing one `tpm_*` command, the `tcsd` daemon will shut down)
If so, could you check with `netstat` if the tcd daemon is listening on port 30003?

Also, could you try to manually execute `chown tss:tss /dev/tpm0 ` in busybox?

**Comment by JorritVrieze** [comment]
Thanks for your quick reply.
I did a full fresh install (clean tpm with jumper), 16.04 i386 desktop and followed your instructions.
For testing i comment out the tpm_nvread -i $INDEX -s 0 >> /dev/null.
Because of this line all next read operations will fail and this creates the 3011 error.
Now i can read the output from tpm_nvread.
I need to restart tcsd after each command but i can see this is working now.
But i still have 2 error in getsecret.sh:  chown not found  and head not found.

When i have a look on the available busybox commands in initramfs and in normal operation i see that a lot of commands are not available in initramfs. Both chown and head are not in the list.
How can i add them to the initramfs ?

**Comment by JorritVrieze** [comment]
I changed the head -c into:  dd bs=$SIZE count=1 and it is working now even with the 'chown' error from the tpm-script.

**Comment by morbitzer** [comment]
This is interesting. I did check again, and in my busybox, I do have the `chown` command as well as the `head` command. I am using Debian jessie amd64. 

The `chown` command is indeed not really necessary - it is more of a safety measure, in case tss is not set to be the owner of `/dev/tpm` for whatever reason. 
Using `dd` instead of `head` is a nice idea. I will do some tests and try to include it in the script in the future, to be sure it also works other versions of busybox.

**Comment by lals1** [comment]
I am receiving the same error in Ubuntu server 16.04 with TPM 1.2. It seems tcsd service in initrd is not starting as it doesn't find the tpm device. I do have tpm and tpm_tis in etc/initramfs-tools/modules.

Any help?

Jun 30 11:40:39 test trousers[1302]:    ...fail!
Jun 30 11:40:39 test unknown[1361]: TCSD TDDL[1361]: TrouSerS ERROR: Could not find a device to open!

Jun 30 11:40:39 test systemd[1]: trousers.service: Control process exited, code=exited status=137
Jun 30 11:40:39 test systemd[1]: Failed to start LSB: starts tcsd.
-- Subject: Unit trousers.service has failed

**Comment by morbitzer** [comment]
Do you have a `/dev/tpm` device under normal operations? 
You could try a `lsmod | grep tpm` to see if your tpm might require any other modules.

**Comment by lals1** [comment]
Yes, here is the output of commands:

root@test:~# ls -l /dev/tpm0
crw------- 1 tss tss 10, 224 Jun 30 16:54 /dev/tpm0

root@test:~# lsmod | grep tpm
tpm_infineon           20480  0

I have now added tpm_infineon in file /etc/initramfs-tools/modules but it still doesn't work.

In busy box, everything works correctly i.e getsecret.sh script, tcsd but its fails in normal mode. I found that trousers service is failing at startup.

root@test:~# systemctl status trousers.service
● trousers.service - LSB: starts tcsd
   Loaded: loaded (/etc/init.d/trousers; bad; vendor preset: enabled)
   Active: failed (Result: exit-code) since Fri 2017-06-30 16:55:18 EEST; 14min ago
     Docs: man:systemd-sysv-generator(8)
  Process: 1389 ExecStart=/etc/init.d/trousers start (code=exited, status=137)

Jun 30 16:55:18 test systemd[1]: Starting LSB: starts tcsd...
Jun 30 16:55:18 test trousers[1389]:  * Starting Trusted Computing daemon tcsd
Jun 30 16:55:18 test trousers[1389]:    ...fail!
Jun 30 16:55:18 test systemd[1]: trousers.service: Control process exited, code=exited status=137
Jun 30 16:55:18 test systemd[1]: Failed to start LSB: starts tcsd.
Jun 30 16:55:18 test systemd[1]: trousers.service: Unit entered failed state.
Jun 30 16:55:18 test systemd[1]: trousers.service: Failed with result 'exit-code'.

And it doesn't get restarted:

root@test:~# systemctl restart trousers.service
Job for trousers.service failed because the control process exited with error code. See "systemctl status trousers.service" and "journalctl -xe" for details.

journalctl -xe output:

-- Unit trousers.service has begun starting up.
Jun 30 17:10:02 test trousers[1656]:  * Starting Trusted Computing daemon tcsd
Jun 30 17:10:02 test tcsd[1662]: TCSD TDDL[1662]: TrouSerS ERROR: Could not find a device to open!
Jun 30 17:10:02 test trousers[1656]:    ...fail!
Jun 30 17:10:02 test systemd[1]: trousers.service: Control process exited, code=exited status=137
Jun 30 17:10:02 test systemd[1]: Failed to start LSB: starts tcsd.


Any clue?

**Comment by morbitzer** [comment]
So the `/dev/tpm0` device is there and the tcsd daemon is running when you enter busybox with `break=init` in the kernel commandline? 
Just to be sure: did you update the initramfs after adding the module?

**Comment by lals1** [comment]
Yes, I can see /dev/tpm0 and tcsd daemon in busybox. /sbin/getsecret.sh script also works.

I have update the initramfs but it still doesn't work. I can't start the trousers service now. It says unable to open the device. Doesn't look like permission issue on tpm device as tss user group is there.

root@test:~# ls -l /dev/tpm0
crw------- 1 tss tss 10, 224 Jun 30 16:54 /dev/tpm0

I tried to uninstall and reinstall trousers package but it is throwing the same error.

**Comment by lals1** [comment]
Ok so apparently I have issues with decrypting the additional disk (USB in my case) from TPM. 
I have tried with just local HDD on a laptop with one partition encrypted and this works quite well.

USB encrypted with LUKS and secret in TPM, can't seem to get the key at the boot time and always ask for passphrase. 

Any tips are welcomed.

**Comment by JorritVrieze** [comment]
Good morning.

 

Can you tell me what i am missing.

 

In 16.04 everything is working fine but in 14.04 when i update the initramfs i get:

 

can not stat ‘/lib/x86_x64-linux-gnu/libns*’  no such file or directory

 

Do you have any idea ?

 

 

Regards,

 

Jorrit

 

Van: morbitzer [mailto:notifications@github.com] 
Verzonden: vrijdag 30 juni 2017 16:28
Aan: fox-it/linux-luks-tpm-boot
CC: JorritVrieze; Comment
Onderwerp: Re: [fox-it/linux-luks-tpm-boot] TPM error during boot (#7)

 

So the /dev/tpm0 device is there and the tcsd daemon is running when you enter busybox with break=init in the kernel commandline?
Just to be sure: did you update the initramfs after adding the module?
